plugin/src/main/java/org/autorefactor/refactoring/rules/SwitchRefactoring.java

```diff
@@ -385,4 +428,99 @@ private Variable extractVariableWithConstantValue(Expression firstOp, Expression
         }
         return null;
     }
+
+    @Override
+    public boolean visit(final SwitchStatement node) {
+        List<Pair<List<Statement>, List<Statement>>> switchStructure = getSwitchStructure(node);
+
+        for (int referenceIndex = 0; referenceIndex < switchStructure.size() - 1; referenceIndex++) {
+            for (int comparedIndex = referenceIndex + 1; comparedIndex < switchStructure.size(); comparedIndex++) {
+                Pair<List<Statement>, List<Statement>> referenceCase = switchStructure.get(referenceIndex);
+                Pair<List<Statement>, List<Statement>> comparedCase = switchStructure.get(comparedIndex);
+
+                if (!referenceCase.getSecond().isEmpty()
+                        && breaksControlFlow(referenceCase.getSecond().get(referenceCase.getSecond().size() - 1))
+                        && isTheSameCode(referenceCase.getSecond(), comparedCase.getSecond())) {
+                    List<Statement> precedingStatements = switchStructure.get(comparedIndex - 1).getSecond();
+
+                    if (breaksControlFlow(precedingStatements.get(precedingStatements.size() - 1))) {
+                        refactorByMergingCases(false, referenceCase, comparedCase);
+                        return DO_NOT_VISIT_SUBTREE;
+                    } else if (referenceIndex == 0) {
+                        refactorByMergingCases(true, comparedCase, referenceCase);
+                        return DO_NOT_VISIT_SUBTREE;
+                    } else {
+                        precedingStatements = switchStructure.get(referenceIndex - 1).getSecond();
+
+                        if (breaksControlFlow(precedingStatements.get(precedingStatements.size() - 1))) {
+                            refactorByMergingCases(true, comparedCase, referenceCase);
+                            return DO_NOT_VISIT_SUBTREE;
+                        }
+                    }
+                }
+            }
+        }
+        return VISIT_SUBTREE;
+    }
+
+    private List<Pair<List<Statement>, List<Statement>>> getSwitchStructure(final SwitchStatement node) {
```

Comment:
I think we are hitting the limits of genericity.
Please consider extracting a class that will replace `Pair<List<Statement>, List<Statement>>`.
This should 1) help with reading the code, maybe move a few methods directly onto this type?